File: Java/src/com/tankbattle/mod/ModLoader.java
package com.tankbattle.mod;

import com.tankbattle.Game;
import java.io.*;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.*;
import java.util.jar.JarFile;

public class ModLoader {

    private static final String MODS_DIR = "mods";
    private static final String MOD_LIST_FILE = "mod_list.txt";
    private static final String GAME_VERSION_FILE = "game_version.txt";
    private static String gameVersion = "1.0.0";

    private static List<ModInfo> mods = new ArrayList<>();
    private static List<IMod> loadedMods = new ArrayList<>();
    private static List<URLClassLoader> classLoaders = new ArrayList<>();

    public static void init() {
        loadGameVersion();

        File modDir = new File(MODS_DIR);
        if (!modDir.exists()) {
            modDir.mkdirs();
            System.out.println("创建 mods 目录");
            createModListFile();
            return;
        }

        createModListFile();
        scanMods();
    }

    private static void createModListFile() {
        File listFile = new File(MODS_DIR, MOD_LIST_FILE);
        if (!listFile.exists()) {
            try (FileWriter fw = new FileWriter(listFile)) {
                fw.write("# Mod list - one mod JAR file per line\n");
                fw.write("# Add # before filename to disable\n");
                fw.write("# Example:\n");
                fw.write("# PlaceholderMod.jar\n");
                fw.write("# #DisabledMod.jar\n");
                fw.write("\n");
                System.out.println("创建模组列表文件: " + MODS_DIR + "/" + MOD_LIST_FILE);
            } catch (IOException e) {
                System.err.println("创建模组列表文件失败: " + e.getMessage());
            }
        }
    }

    private static void loadGameVersion() {
        File versionFile = new File(GAME_VERSION_FILE);
        if (versionFile.exists()) {
            try (BufferedReader br = new BufferedReader(new FileReader(versionFile))) {
                gameVersion = br.readLine().trim();
                System.out.println("游戏版本: " + gameVersion);
            } catch (IOException e) {
                System.err.println("读取游戏版本失败，使用默认值: " + gameVersion);
            }
        } else {
            try (FileWriter fw = new FileWriter(versionFile)) {
                fw.write(gameVersion);
                System.out.println("创建游戏版本文件: " + GAME_VERSION_FILE);
            } catch (IOException e) {
                System.err.println("创建版本文件失败");
            }
        }
    }

    private static void scanMods() {
        File modDir = new File(MODS_DIR);
        File[] jars = modDir.listFiles((dir, name) -> name.endsWith(".jar"));

        if (jars == null || jars.length == 0) {
            System.out.println("未找到模组文件");
            return;
        }

        // 获取所有模组文件名（包括被禁用的）
        Set<String> allModFiles = new HashSet<>();
        for (File jar : jars) {
            allModFiles.add(jar.getName());
        }

        // 获取启用的模组列表
        Set<String> enabledMods = loadEnabledMods();

        System.out.println("发现 " + jars.length + " 个模组文件");

        // 先处理所有 JAR 文件
        Map<String, File> jarMap = new HashMap<>();
        for (File jar : jars) {
            jarMap.put(jar.getName(), jar);
        }

        // 处理所有模组文件
        for (String jarName : allModFiles) {
            File jarFile = jarMap.get(jarName);
            boolean isEnabled = enabledMods.contains(jarName);

            if (!isEnabled) {
                // 模组被禁用，显示为禁用状态
                ModInfo info = new ModInfo(jarName.replace(".jar", ""), "?", "?", "");
                info.setLoaded(false);
                info.setError("已禁用");
                mods.add(info);
                System.out.println("跳过禁用模组: " + jarName);
            } else {
                loadMod(jarFile);
            }
        }
    }

    private static Set<String> loadEnabledMods() {
        Set<String> enabled = new HashSet<>();
        File listFile = new File(MODS_DIR, MOD_LIST_FILE);

        if (!listFile.exists()) {
            return enabled;
        }

        try (BufferedReader br = new BufferedReader(new FileReader(listFile))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) continue;
                if (line.startsWith("#")) continue;

                if (line.endsWith(".jar")) {
                    enabled.add(line);
                }
            }
        } catch (IOException e) {
            System.err.println("读取模组列表失败: " + e.getMessage());
        }

        return enabled;
    }

    private static void loadMod(File jarFile) {
        try {
            URLClassLoader loader = new URLClassLoader(
                new URL[]{jarFile.toURI().toURL()},
                Thread.currentThread().getContextClassLoader()
            );

            java.util.jar.Manifest manifest = getManifest(jarFile);
            if (manifest == null) {
                System.err.println("警告 " + jarFile.getName() + " 缺少 MANIFEST.MF");
                ModInfo info = new ModInfo(jarFile.getName().replace(".jar", ""), "?", "?", "");
                info.setLoaded(false);
                info.setError("缺少 MANIFEST.MF");
                mods.add(info);
                return;
            }

            Map<String, String> attrs = new HashMap<>();
            manifest.getMainAttributes().forEach((key, value) -> {
                attrs.put(key.toString(), value.toString());
            });

            String modName = attrs.getOrDefault("Mod-Name", jarFile.getName().replace(".jar", ""));
            String modVersion = attrs.getOrDefault("Mod-Version", "1.0.0");
            String requiredGameVersion = attrs.getOrDefault("Game-Version", "1.0.0");
            String mainClass = attrs.get("Main-Class");

            if (mainClass == null) {
                System.err.println("警告 " + jarFile.getName() + " 缺少 Main-Class");
                ModInfo info = new ModInfo(modName, modVersion, requiredGameVersion, mainClass);
                info.setLoaded(false);
                info.setError("缺少 Main-Class");
                mods.add(info);
                return;
            }

            ModInfo info = new ModInfo(modName, modVersion, requiredGameVersion, mainClass);
            mods.add(info);

            if (!isCompatible(requiredGameVersion)) {
                String error = "需要游戏 v" + requiredGameVersion + "，当前 v" + gameVersion;
                info.setError(error);
                System.err.println("错误 " + modName + ": " + error);
                return;
            }

            Class<?> clazz = loader.loadClass(mainClass);
            if (!IMod.class.isAssignableFrom(clazz)) {
                String error = "未实现 IMod 接口";
                info.setError(error);
                System.err.println("错误 " + modName + ": " + error);
                return;
            }

            IMod mod = (IMod) clazz.getDeclaredConstructor().newInstance();
            info.setLoaded(true);
            loadedMods.add(mod);
            classLoaders.add(loader);
            System.out.println("加载模组: " + modName + " v" + modVersion);

        } catch (Exception e) {
            ModInfo info = new ModInfo(jarFile.getName().replace(".jar", ""), "?", "?", "");
            info.setLoaded(false);
            info.setError("加载失败: " + e.getMessage());
            mods.add(info);
            System.err.println("错误 " + jarFile.getName() + " 加载失败: " + e.getMessage());
        }
    }

    private static java.util.jar.Manifest getManifest(File jarFile) {
        try (JarFile jar = new JarFile(jarFile)) {
            return jar.getManifest();
        } catch (IOException e) {
            return null;
        }
    }

    private static boolean isCompatible(String required) {
        if (required.equals("*") || required.equals(gameVersion)) return true;

        try {
            String[] reqParts = required.split("\\.");
            String[] gameParts = gameVersion.split("\\.");
            int minLen = Math.min(reqParts.length, gameParts.length);

            for (int i = 0; i < minLen; i++) {
                int req = Integer.parseInt(reqParts[i]);
                int game = Integer.parseInt(gameParts[i]);
                if (game != req) {
                    return game > req;
                }
            }
            return reqParts.length <= gameParts.length;
        } catch (NumberFormatException e) {
            return required.equals(gameVersion) || required.equals("*");
        }
    }

    public static void onGameLoad(Game game) {
        for (IMod mod : loadedMods) {
            try {
                mod.onLoad(game);
                System.out.println("模组初始化完成: " + mod.getModName());
            } catch (Exception e) {
                System.err.println("模组 " + mod.getModName() + " 初始化失败: " + e.getMessage());
            }
        }
    }

    public static void onGameUpdate(double dt) {
        for (IMod mod : loadedMods) {
            try {
                mod.onUpdate(dt);
            } catch (Exception e) {
            }
        }
    }

    public static void unloadAll() {
        for (IMod mod : loadedMods) {
            try {
                mod.onUnload();
            } catch (Exception e) {
            }
        }

        for (URLClassLoader loader : classLoaders) {
            try {
                loader.close();
            } catch (IOException e) {
            }
        }
        classLoaders.clear();
        loadedMods.clear();
        mods.clear();
    }

    public static List<ModInfo> getMods() {
        return mods;
    }

    public static List<IMod> getLoadedMods() {
        return loadedMods;
    }

    public static String getGameVersion() {
        return gameVersion;
    }

    public static void toggleMod(int index) {
        if (index < 0 || index >= mods.size()) return;

        ModInfo info = mods.get(index);
    
        // 尝试匹配 jar 文件名
        String jarName = info.getName();
        if (!jarName.endsWith(".jar")) {
            jarName = jarName + ".jar";
        }

        File listFile = new File(MODS_DIR, MOD_LIST_FILE);
        List<String> lines = new ArrayList<>();
        boolean found = false;

        try (BufferedReader br = new BufferedReader(new FileReader(listFile))) {
            String line;
            while ((line = br.readLine()) != null) {
                String trimmed = line.trim();
                String clean = trimmed.replaceFirst("^#\\s*", "").trim();

                // 匹配时忽略大小写
                if (clean.equalsIgnoreCase(jarName) || clean.equalsIgnoreCase(info.getName())) {
                    found = true;
                    if (line.trim().startsWith("#")) {
                        lines.add(clean);
                    } else {
                        lines.add("# " + clean);
                    }
                } else {
                    lines.add(line);
                }
            }
        } catch (IOException e) {
            System.err.println("读取模组列表失败: " + e.getMessage());
            return;
        }

        if (!found) {
            if (info.isLoaded()) {
                lines.add("# " + jarName);
            } else {
                lines.add(jarName);
            }
        }

        try (FileWriter fw = new FileWriter(listFile)) {
            for (String line : lines) {
                fw.write(line + "\n");
            }
        } catch (IOException e) {
            System.err.println("写入模组列表失败: " + e.getMessage());
            return;
        }

        System.out.println("模组切换: " + info.getName() + " -> " + 
            (info.isLoaded() ? "已禁用 (重启生效)" : "已启用 (重启生效)"));
    }
}

File: C++/src/mod/ModLoader.cpp
// src/mod/ModLoader.cpp
#include "ModLoader.h"
#include "../Game.h"
#include <iostream>
#include <fstream>
#include <filesystem>
#include <algorithm>

#ifdef _WIN32
#include <windows.h>
#else
#include <dlfcn.h>
#endif

namespace fs = std::filesystem;

ModLoader& ModLoader::getInstance() {
    static ModLoader instance;
    return instance;
}

ModLoader::ModLoader()
    : modsDir("mods"), listFile("mods/mod_list.txt"),
      versionFile("game_version.txt"), gameVersion("1.0.0"), initialized(false) {
    loadGameVersion();
}

ModLoader::~ModLoader() {
    unloadAll();
}

void ModLoader::loadGameVersion() {
    std::ifstream file(versionFile);
    if (file.is_open()) {
        std::string line;
        if (std::getline(file, line)) {
            while (!line.empty() && (line.back() == ' ' || line.back() == '\t' || line.back() == '\r')) {
                line.pop_back();
            }
            while (!line.empty() && (line.front() == ' ' || line.front() == '\t')) {
                line.erase(0, 1);
            }
            if (!line.empty()) {
                gameVersion = line;
                std::cout << "游戏版本: " << gameVersion << " (从 " << versionFile << ")" << std::endl;
            }
        }
        file.close();
    } else {
        saveGameVersion();
        std::cout << "创建游戏版本文件: " << versionFile << std::endl;
    }
}

void ModLoader::saveGameVersion() {
    std::ofstream file(versionFile);
    if (file.is_open()) {
        file << gameVersion << "\n";
        file.close();
        std::cout << "保存游戏版本: " << gameVersion << " -> " << versionFile << std::endl;
    } else {
        std::cerr << "保存游戏版本失败: " << versionFile << std::endl;
    }
}

void ModLoader::init() {
    if (initialized) return;

    if (gameVersion.empty()) {
        loadGameVersion();
    }

    if (!fs::exists(modsDir)) {
        fs::create_directories(modsDir);
        std::cout << "创建 mods 目录" << std::endl;
    }

    if (!fs::exists(listFile)) {
        std::ofstream file(listFile);
        if (file.is_open()) {
            file << "# Mod list - one mod DLL/SO file per line\n";
            file << "# Add # before filename to disable\n";
            file << "# Example:\n";
            file << "# MyMod.dll\n";
            file << "# #DisabledMod.dll\n";
            file.close();
            std::cout << "创建模组列表文件: " << listFile << std::endl;
        }
    }

    loadEnabledList();
    scanMods();
    initialized = true;
    std::cout << "模组系统初始化完成 (游戏版本: " << gameVersion << ")" << std::endl;
}

void ModLoader::loadEnabledList() {
    enabledMods.clear();
    std::ifstream file(listFile);
    if (!file.is_open()) return;

    std::string line;
    while (std::getline(file, line)) {
        size_t start = line.find_first_not_of(" \t\r\n");
        if (start == std::string::npos) continue;
        size_t end = line.find_last_not_of(" \t\r\n");
        line = line.substr(start, end - start + 1);

        if (line.empty() || line[0] == '#') continue;
        enabledMods.push_back(line);
    }
    file.close();
}

void ModLoader::saveEnabledList() {
    std::ofstream file(listFile);
    if (!file.is_open()) return;

    file << "# Mod list - one mod DLL/SO file per line\n";
    file << "# Add # before filename to disable\n\n";

    for (const auto& mod : mods) {
        if (mod.disabled) {
            file << "# " << mod.filePath << "\n";
        } else {
            file << mod.filePath << "\n";
        }
    }
    file.close();
}

void ModLoader::scanMods() {
    mods.clear();

    for (const auto& entry : fs::directory_iterator(modsDir)) {
        if (!entry.is_regular_file()) continue;

        std::string path = entry.path().string();
        std::string filename = entry.path().filename().string();

#ifdef _WIN32
        if (filename.find(".dll") == std::string::npos) continue;
#else
        if (filename.find(".so") == std::string::npos) continue;
#endif

        bool isEnabled = std::find(enabledMods.begin(), enabledMods.end(), filename) != enabledMods.end();

        if (!isEnabled) {
            ModInfo info(filename, "?", "?", filename);
            info.disabled = true;
            mods.push_back(info);
            continue;
        }

        loadMod(path);
    }

    std::cout << "发现 " << mods.size() << " 个模组" << std::endl;
}

bool ModLoader::loadMod(const std::string& filePath) {
    fs::path path(filePath);
    std::string filename = path.filename().string();

#ifdef _WIN32
    HMODULE handle = LoadLibraryA(filePath.c_str());
    if (!handle) {
        ModInfo info(filename, "?", "?", filename);
        info.error = "加载失败: " + std::to_string(GetLastError());
        mods.push_back(info);
        std::cerr << "错误 " << filename << " 加载失败" << std::endl;
        return false;
    }

    auto createFunc = (CreateModFunc)GetProcAddress(handle, "createMod");
    auto destroyFunc = (DestroyModFunc)GetProcAddress(handle, "destroyMod");

    if (!createFunc || !destroyFunc) {
        FreeLibrary(handle);
        ModInfo info(filename, "?", "?", filename);
        info.error = "缺少 createMod/destroyMod 导出函数";
        mods.push_back(info);
        std::cerr << "错误 " << filename << " 缺少导出函数" << std::endl;
        return false;
    }
#else
    void* handle = dlopen(filePath.c_str(), RTLD_LAZY);
    if (!handle) {
        ModInfo info(filename, "?", "?", filename);
        info.error = "加载失败: " + std::string(dlerror());
        mods.push_back(info);
        std::cerr << "错误 " << filename << " 加载失败" << std::endl;
        return false;
    }

    auto createFunc = (CreateModFunc)dlsym(handle, "createMod");
    auto destroyFunc = (DestroyModFunc)dlsym(handle, "destroyMod");

    if (!createFunc || !destroyFunc) {
        dlclose(handle);
        ModInfo info(filename, "?", "?", filename);
        info.error = "缺少 createMod/destroyMod 导出函数";
        mods.push_back(info);
        std::cerr << "错误 " << filename << " 缺少导出函数" << std::endl;
        return false;
    }
#endif

    IMod* instance = createFunc();
    if (!instance) {
#ifdef _WIN32
        FreeLibrary(handle);
#else
        dlclose(handle);
#endif
        ModInfo info(filename, "?", "?", filename);
        info.error = "创建模组实例失败";
        mods.push_back(info);
        return false;
    }

    std::string modName = instance->getModName();
    std::string modVersion = instance->getModVersion();
    std::string requiredVersion = instance->getGameVersion();

    if (!isCompatible(requiredVersion)) {
        instance->onUnload();
        destroyFunc(instance);
#ifdef _WIN32
        FreeLibrary(handle);
#else
        dlclose(handle);
#endif
        ModInfo info(modName, modVersion, requiredVersion, filename);
        info.error = "需要游戏 v" + requiredVersion + "，当前 v" + gameVersion;
        mods.push_back(info);
        std::cerr << "错误 " << modName << ": " << info.error << std::endl;
        return false;
    }

    ModInfo info(modName, modVersion, requiredVersion, filename);
    info.loaded = true;
    mods.push_back(info);

    LoadedModule mod;
    mod.handle = handle;
    mod.instance = instance;
    mod.createFunc = createFunc;
    mod.destroyFunc = destroyFunc;
    mod.filePath = filename;
    loadedModules.push_back(mod);
    loadedMods.push_back(instance);

    std::cout << "加载模组: " << modName << " v" << modVersion << " (需要游戏 " << requiredVersion << ")" << std::endl;
    return true;
}

void ModLoader::unloadMod(int index) {
    if (index < 0 || index >= (int)loadedModules.size()) return;

    auto& mod = loadedModules[index];
    if (mod.instance) {
        mod.instance->onUnload();
        mod.destroyFunc(mod.instance);
    }

#ifdef _WIN32
    FreeLibrary((HMODULE)mod.handle);
#else
    dlclose(mod.handle);
#endif

    loadedModules.erase(loadedModules.begin() + index);
    loadedMods.erase(loadedMods.begin() + index);
}

void ModLoader::unloadAll() {
    for (int i = (int)loadedModules.size() - 1; i >= 0; --i) {
        unloadMod(i);
    }
    loadedMods.clear();
    loadedModules.clear();
}

void ModLoader::onGameLoad(Game* game) {
    for (auto* mod : loadedMods) {
        try {
            mod->onLoad(game);
            std::cout << "模组初始化完成: " << mod->getModName() << std::endl;
        } catch (const std::exception& e) {
            std::cerr << "模组 " << mod->getModName() << " 初始化失败: " << e.what() << std::endl;
        }
    }
}

void ModLoader::onGameUpdate(float dt) {
    for (auto* mod : loadedMods) {
        try {
            mod->onUpdate(dt);
        } catch (const std::exception&) {
        }
    }
}

void ModLoader::toggleMod(int index) {
    if (index < 0 || index >= (int)mods.size()) return;

    auto& info = mods[index];
    info.disabled = !info.disabled;

    if (info.loaded) {
        for (int i = 0; i < (int)loadedModules.size(); ++i) {
            if (loadedModules[i].filePath == info.filePath) {
                unloadMod(i);
                break;
            }
        }
        info.loaded = false;
    }

    saveEnabledList();

    std::cout << "模组切换: " << info.name << " -> "
              << (info.disabled ? "已禁用" : "已启用") << " (重启生效)" << std::endl;
}

void ModLoader::reloadAll() {
    unloadAll();
    mods.clear();
    loadEnabledList();
    scanMods();
}

bool ModLoader::isCompatible(const std::string& required) {
    if (required == "*") return true;
    if (required == gameVersion) return true;

    try {
        std::string req = required;
        std::string ver = gameVersion;

        if (req[0] == 'v' || req[0] == 'V') req = req.substr(1);
        if (ver[0] == 'v' || ver[0] == 'V') ver = ver.substr(1);

        auto split = [](const std::string& s) {
            std::vector<int> parts;
            std::string current;
            for (char c : s) {
                if (c == '.') {
                    if (!current.empty()) {
                        parts.push_back(std::stoi(current));
                        current.clear();
                    }
                } else if (isdigit(c)) {
                    current += c;
                }
            }
            if (!current.empty()) {
                parts.push_back(std::stoi(current));
            }
            return parts;
        };

        std::vector<int> reqParts = split(req);
        std::vector<int> verParts = split(ver);

        size_t minLen = std::min(reqParts.size(), verParts.size());
        for (size_t i = 0; i < minLen; ++i) {
            if (reqParts[i] != verParts[i]) {
                return reqParts[i] < verParts[i];
            }
        }

        return reqParts.size() <= verParts.size();

    } catch (const std::exception&) {
        return required == gameVersion || required == "*";
    }
}